ui/src/message_popup/scaffolding_card.jsx

```diff
@@ -1,61 +1,149 @@
-    helpType: React.PropTypes.string.isRequired,
-    onHelpToggled: React.PropTypes.func.isRequired,
-    itemsToShow: React.PropTypes.object.isRequired
+    scaffolding: React.PropTypes.shape({
+      helpType: React.PropTypes.string.isRequired,
+      shouldShowStudentCard: React.PropTypes.bool.isRequired,
+      shouldShowSummary: React.PropTypes.bool.isRequired
+    }).isRequired,
+    email: React.PropTypes.string.isRequired,
+    itemsToShow: React.PropTypes.object.isRequired,
+    saveScaffoldingAndSession: React.PropTypes.func.isRequired
   },
   
+  getInitialState(){
+    const isSolutionMode = _.has(this.props.itemsToShow, 'solution');
+    return ({
+      email: this.props.email,
+      questions: allQuestions,
+      sessionLength: 10,
+      helpType: this.props.scaffolding.helpType,
+      shouldShowStudentCard: this.props.scaffolding.shouldShowStudentCard,
+      shouldShowSummary: this.props.scaffolding.shouldShowSummary,
+      competencyGroupValue: ALL_COMPETENCY_GROUPS,
+      isSolutionMode
+    });
+  },
+
+  getQuestions(competencyGroup=""){
+    var {competencyGroupValue, isSolutionMode} = this.state;
+    if(competencyGroup !== ""){
+      competencyGroupValue = competencyGroup;
+    }
+    const questions = (isSolutionMode || competencyGroupValue === ALL_COMPETENCY_GROUPS)
+      ? _.shuffle(this.withStudents(allQuestions))
+      : this.questionsForCompetencies(competencyGroupValue);
+    return questions;
+  },
+  
+  withStudents(questions) {
+    return questions.map((question) => {
+      const student = _.find(allStudents, {id: question.studentId });
+      return _.extend({student}, question);
+    });
+  },
+  
+  questionsForCompetencies(competencyGroup) {
+    const withCompetencyGroups = _.compact(allQuestions.map((question) => {
+      const learningObjective = _.find(learningObjectives, { id: question.learningObjectiveId });
+      if (learningObjective.competencyGroup !== competencyGroup) return null;
+      return {
+        ...question,
+        competencyGroup: learningObjective.competencyGroup
+      };
+    }));
+    return _.shuffle(this.withStudents(withCompetencyGroups));
+  },
+
+  onSave(){
+    const scaffolding = {
+      helpType: this.state.helpType,
+      shouldShowStudentCard: this.state.shouldShowStudentCard,
+      shouldShowSummary: this.state.shouldShowSummary
+    };
+    const gameSession = {
+      email: this.state.email,
+      sessionLength: this.state.sessionLength,
+      questions: this.getQuestions(this.state.competencyGroupValue)
+    };
+    this.props.saveScaffoldingAndSession(scaffolding, gameSession);
+  },
   
+  onCompetencyGroupChanged(event, competencyGroupValue){
+    const questions = this.getQuestions(competencyGroupValue);
+    const newLength = questions.length;
+    var sessionLength = this.state.sessionLength;
+    if(sessionLength > newLength){
+      sessionLength = newLength;
+    }
+    this.setState({questions, sessionLength, competencyGroupValue});
```

Comment:
I don't think we need to save the `questions` as state - they can be computed from the other state values when we need it (like in `onSave`).